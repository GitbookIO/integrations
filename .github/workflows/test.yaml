name: Test

on:
    push:

jobs:
    test:
        runs-on: ubuntu-latest
        strategy:
            matrix:
                node-version: [18.x]
        steps:
            - uses: actions/checkout@v3
            - name: Use Node.js ${{ matrix.node-version }}
              uses: actions/setup-node@v3
              with:
                  node-version: ${{ matrix.node-version }}
            - name: Install Dependencies
              run: npm ci
            - name: Run tests
              run: npm run test
    lint:
        runs-on: ubuntu-latest
        strategy:
            matrix:
                node-version: [18.x]
        steps:
            - uses: actions/checkout@v3
            - name: Use Node.js ${{ matrix.node-version }}
              uses: actions/setup-node@v3
              with:
                  node-version: ${{ matrix.node-version }}
            - name: Install Dependencies
              run: npm ci
            - name: Lint
              run: npm run lint
