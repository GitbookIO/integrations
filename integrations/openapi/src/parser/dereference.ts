import { ExposableError } from '@gitbook/runtime';
import { dereference, type OpenAPIV3 } from '@gitbook/openapi-parser';

/**
 * Dereference the OpenAPI and return the schema.
 */
export async function getDereferencedSchema(
    args: Parameters<typeof dereference>['0'],
): Promise<OpenAPIV3.Document> {
    const dereferenceResult = await dereference(args);

    if (!dereferenceResult.schema) {
        throw new ExposableError('Failed to dereference OpenAPI spec');
    }

    return dereferenceResult.schema as OpenAPIV3.Document;
}
