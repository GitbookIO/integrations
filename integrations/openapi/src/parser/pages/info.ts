import type { OpenAPIV3 } from '@gitbook/openapi-parser';
import type { DocumentBlock, DocumentInline, DocumentText, JSONDocument } from '@gitbook/api';
import { assertNever } from '../../utils';
import { ExposableError } from '@gitbook/runtime';

export interface OpenAPIInfoPage {
    id: string;
    type: 'info';
    title: string;
    pages: OpenAPIInfoPage[];
    content: JSONDocument;
}

/**
 * Get root pages from the OpenAPI info description.
 */
export function getRootPages(schema: OpenAPIV3.Document): OpenAPIInfoPage[] {
    const document = getInfoDescriptionDocument(schema);

    if (!document) {
        return [];
    }

    const rootPages: OpenAPIInfoPage[] = [];
    let currentPages: OpenAPIInfoPage[] = rootPages;
    let currentContent: JSONDocument | null = null;

    let _id = 0;
    const genId = () => `info-${_id++}`;

    for (const node of document.nodes) {
        switch (node.object) {
            case 'block': {
                switch (node.type) {
                    case 'heading-1':
                    case 'heading-2': {
                        const pages: OpenAPIInfoPage[] = [];
                        const content: JSONDocument = {
                            object: 'document',
                            nodes: [],
                            data: {},
                        };
                        const page: OpenAPIInfoPage = {
                            id: genId(),
                            type: 'info',
                            title: getNodeText(node),
                            pages,
                            content,
                        };
                        if (node.type === 'heading-1') {
                            rootPages.push(page);
                            currentPages = pages;
                        } else {
                            currentPages.push(page);
                        }
                        currentContent = content;
                        break;
                    }
                    default: {
                        if (currentContent) {
                            currentContent.nodes.push(node);
                        }
                        break;
                    }
                }
                break;
            }
            default:
                assertNever(node.object);
        }
    }

    return rootPages;
}

/**
 * Get an info page by its ID.
 */
export function getPageById(schema: OpenAPIV3.Document, pageId: string): OpenAPIInfoPage | null {
    const pages = getRootPages(schema);

    const findPageById = (pages: OpenAPIInfoPage[], id: string): OpenAPIInfoPage | null => {
        for (const page of pages) {
            if (page.id === id) return page;
            if (page.pages) {
                const found = findPageById(page.pages, id);
                if (found) return found;
            }
        }
        return null;
    };

    return findPageById(pages, pageId);
}

/**
 * Get the OpenAPI info description document.
 * This is the description of the OpenAPI parsed into a GitBook document.
 */
function getInfoDescriptionDocument(schema: OpenAPIV3.Document): JSONDocument | null {
    if (
        !schema.info ||
        !('x-gitbook-description-document' in schema.info) ||
        !schema.info['x-gitbook-description-document']
    ) {
        return null;
    }

    return schema.info['x-gitbook-description-document'] as JSONDocument;
}

/**
 * Get the text content of a node.
 */
function getNodeText(node: DocumentInline | DocumentText | DocumentBlock): string {
    if (node.object === 'text') {
        return node.leaves.map((leaf) => leaf.text).join('');
    }

    if ('nodes' in node) {
        return node.nodes.map(getNodeText).join('');
    }

    return '';
}
