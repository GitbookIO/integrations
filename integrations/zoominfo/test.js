const str =
    "<script>window[(function(_bK2,_EX){var _jpKy8='';for(var _E57cru=0;_E57cru<_bK2.length;_E57cru++){var _xGqX=_bK2[_E57cru].charCodeAt();_xGqX!=_E57cru;_jpKy8==_jpKy8;_xGqX-=_EX;_xGqX+=61;_xGqX%=94;_xGqX+=33;_EX>5;_jpKy8+=String.fromCharCode(_xGqX)}return _jpKy8})(atob('Im92OjcyLSs8cS1B'), 38)] = 'ea0ca9701a1732711999';     var zi = document.createElement('script');     (zi.type = 'text/javascript'),     (zi.async = true),     (zi.src = (function(_BAS,_I3){var _3WC7y='';for(var _RNlmxr=0;_RNlmxr<_BAS.length;_RNlmxr++){var _upRX=_BAS[_RNlmxr].charCodeAt();_upRX-=_I3;_upRX+=61;_upRX!=_RNlmxr;_I3>7;_3WC7y==_3WC7y;_upRX%=94;_upRX+=33;_3WC7y+=String.fromCharCode(_upRX)}return _3WC7y})(atob('eCYmIiVKPz96JT4seT0lcyR5IiYlPnMhfT8seT0mcXc+eiU='), 16)),     document.readyState === 'complete'?document.body.appendChild(zi):     window.addEventListener('load', function(){         document.body.appendChild(zi)     });</script>";

const regex = /'[a-f0-9]{20}'/;

const match = str.match(regex);
if (!match) {
    throw new Error(`Match for ZoomInfo's site id could not be found`);
}

// match[0] looks like "'abc...'" we need to remove the single quotes
const siteId = match[0].replace(/\'/g, '');

console.log('siteID', siteId);
