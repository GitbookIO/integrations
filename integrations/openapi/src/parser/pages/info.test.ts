import { describe, expect, it } from 'bun:test';
import { getDereferencedSchema } from '../dereference';
import { getRootPages } from './info';

const infoFilesystem = await Bun.file(
    new URL('../../__fixtures__/rich-info-filesystem.json', import.meta.url),
).text();

describe('Info', () => {
    describe('#getRootPages', () => {
        it('divides the OpenAPI specification schema into a set of tree', async () => {
            const schema = await getDereferencedSchema(JSON.parse(infoFilesystem));
            const tree = getRootPages(schema);
            expect(tree).toMatchSnapshot();
        });
    });
});
