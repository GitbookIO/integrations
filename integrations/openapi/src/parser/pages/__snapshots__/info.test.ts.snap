// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Info #getRootPages divides the OpenAPI specification schema into a set of tree 1`] = `
[
  {
    "content": {
      "data": {},
      "nodes": [
        {
          "data": {},
          "isVoid": false,
          "nodes": [
            {
              "leaves": [
                {
                  "marks": [],
                  "object": "leaf",
                  "text": "This is great!",
                },
              ],
              "object": "text",
            },
          ],
          "object": "block",
          "type": "paragraph",
        },
      ],
      "object": "document",
    },
    "id": "info-0",
    "pages": [
      {
        "content": {
          "data": {},
          "nodes": [
            {
              "data": {},
              "isVoid": false,
              "nodes": [
                {
                  "leaves": [
                    {
                      "marks": [],
                      "object": "leaf",
                      "text": "This is the petstore!",
                    },
                  ],
                  "object": "text",
                },
              ],
              "object": "block",
              "type": "paragraph",
            },
            {
              "data": {
                "style": "info",
              },
              "isVoid": false,
              "nodes": [
                {
                  "data": {},
                  "isVoid": false,
                  "nodes": [
                    {
                      "leaves": [
                        {
                          "marks": [],
                          "object": "leaf",
                          "text": "We have a beautiful hint here",
                        },
                      ],
                      "object": "text",
                    },
                  ],
                  "object": "block",
                  "type": "paragraph",
                },
              ],
              "object": "block",
              "type": "hint",
            },
          ],
          "object": "document",
        },
        "id": "info-1",
        "pages": [],
        "title": "About",
        "type": "info",
      },
    ],
    "title": "Getting started now",
    "type": "info",
  },
]
`;
